Add Issue 1: Introduction to Snakefiles (#15)
USGS-R/snakemake-tutorial-scratch: issue_drafts/issue_1.md

# Issue 1: Introduction to Snakefiles

Pipeline steps: Download the temperature predictions.
Concepts learned: Snakefile structure, the `script` directive, basic `snakemake` commands 

## Tutorial Overview

Welcome to the USGS Data Science Practitioner's Snakemake tutorial!
This tutorial consists of several issues.
Each issue will guide you in building a piece of a Snakemake data pipeline.
This pipeline will download, aggregate, and visualize predictions of lake temperatures at multiple depths.
At the same time, each issue will introduce you to new Snakemake concepts.

Most of the Python code for the pipeline is already written.
It's organized into three phases.
- `1_fetch` downloads the predictions from this data release on ScienceBase: https://www.sciencebase.gov/catalog/item/5e5d0bb9e4b01d50924f2b36 and unzips the files.
- `2_process` computes the mean prediction on each day of the year for a selection of lakes and aggregates those means into a single .csv file.
- `3_plot` plots the mean temperatures on each day of the year for each lake at a selection of depths.

> You can find the code organized by phase in three folders.

Reviewer: ```suggestion
The code for each of the phases is located in the correspondingly named folder of this repository.
```

Within each folder are all the Python scripts needed to execute the tasks of that phase. 
If we wanted to carry out these tasks using the scripts as they are, we could execute each script in order.
Alternatively, we can use a Snakemake pipeline.
Snakemake allows us to:

> - Automate the execution of the scripts

Reviewer: What do you think of adding this bullet? I know you can always just write documentation to describe the order of your workflow and how it all fits together, but I think one of the benefits to using a pipelining tool is having this formally documented. 
```suggestion
- Describe the order in which scripts should run, and what the expected inputs/outputs of each step are
- Automate the execution of the scripts
```

- Track which outputs have already been made
- Only run the parts of the pipeline that need to be run
- Rerun parts of the pipeline, or the entire pipeline at once
- Easily expand the pipeline to accommodate more data.

In this issue we'll create a Snakefile and write our first rule.

## What's a Snakefile?

A Snakefile is a file that describes every step a pipeline will take, which files that step depends upon, and which files that step produces.
The Snakefile can also hold or point to configuration settings.
It's like a recipe for how to create outputs, or a blueprint of the pipeline.

> The main pieces of a Snakefile are called rules.

Reviewer: ```suggestion
A Snakefile is broken down into a sequential set of steps, each of which is called a rule.
```

## What's a rule?

A rule describes one piece of a pipeline - one step that it can take.
A rule specifies:
- the files that are inputs to that step
- the files that are outputs of that step
- how to execute the step
- parameters and settings that pertain to that step

## Running the script without the pipeline

> This issue's rule will download the temperature predictions from ScienceBase that the pipeline will process and visualize.

Reviewer: ```suggestion
The rule that we will create in this issue will download the temperature predictions from ScienceBase that we will process and visualize in later steps.
```

As mentioned earlier, the Python code for this step is already written.
You can find it in the script `sb_get.py`.

> Try running that script now. I'll wait!
> ```
> python 1_fetch/sb_get.py
> ```
> Be sure that you're in the correct Conda environment - if not, activate the right environment first:
> ```
> conda activate snakemake-tutorial
> ```

Reviewer: ```suggestion
Let's test out the Python code before we start to build our pipeline. Navigate to your local clone of this repository in either Anaconda Navigator or the command line. Be sure that you're in the Conda environment we created for this tutorial - if not, activate the right environment first:
```
conda activate snakemake-tutorial
```

Now, try running the script. I'll wait!
```
python 1_fetch/sb_get.py
```
```

The script should download a single zip file to `1_fetch/out/pgdl_predictions_04_N45.50-48.00_W92.00-93.00.zip`.
If the file wasn't downloaded, let your instructor know and we'll get it sorted out!

If the file is there, great!

> Now, let's write a Snakemake rule so that we build in the step of downloading that file to the pipeline.

Reviewer: ```suggestion
Now, let's write a Snakemake rule to download that file and build the first step of our pipeline!
```

There are a few ways to do this, but we'll start with our favorite way: the `script` directive.
The `script` directive tells Snakemake to run a script in order to make the output file(s) for this rule.
In this case, we know that all we need to do is to run `1_fetch/sb_get.py` to download the file.
So, let's specify that Python file as our script.
```
rule get_sb_data:
    output:
        1_fetch/out/pgdl_predictions_04_N45.50-48.00_W92.00-93.00.zip
    script:
        1_fetch/sb_get.py
```
Like with any Python file, be sure to pay attention to indentation!
Since the script is a `.py`, Snakemake knows to use `python` to run it.
Unless we specify otherwise, the version of Python that's used to run the script will be the version that's called when you type `python` into the command line.
So, when you want to run the pipeline, be sure that you're in the correct Conda environment.

Okay, we've got our Snakefile written with its first rule!
Let's try it out.

## Executing the pipeline

First, go ahead and delete the zip file we downloaded earlier - the one that's in `1_fetch/out/`.
Next, execute our new rule using the following command:
```

> snakemake --cores 1 1_fetch/out/pgdl_predictions_04_N45.50-48.00_W92.00-93.00.zip

Reviewer: I honestly haven't done much rule-by-rule running of snakemake. It seems easier to type the rule name than to remember what the output file would be and type it. Should we suggest that instead? I don't have a hard stance...I'm curious if you think one is better than the other.
```suggestion
snakemake --cores 1 -R get_sb_data
```

Author: If a rule's output has wildcards in it, then the only way to run that rule is to specify its output. For that reason, I've become accustomed to specifying output files instead of rule names. Using the output file as an argument works in most if not all cases. My leaning is to demonstrate this practice first so that folks "imprint" on it.

```
This is the main way to run the pipeline.
We call the `snakemake` executable, passing in the file or files we want it to create as arguments.
Notice the `--cores 1` argument.

> That specifies how many cores we want to use.

Reviewer: ```suggestion
That specifies how many cores we want to use, and it is a required argument that you must pass each time you run snakemake.
```
